bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,224 @@
+  end
+
+  opts.on('--log-sql',
+    'Logs all sql statements sent to the database'
+  ) do |log_sql|
+    options[:log_sql] = log_sql
+  end
+
+  opts.on('--[no-]slack',
+    'Sends summary of script to slack. Defaults to true.'
+  ) do |notify_slack|
+    options[:notify_slack] = notify_slack
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Initial parsed options: #{options}"
+
+PEGASUS_DB.loggers << CDO.log if options[:log_sql]
+
+if options[:start_date].nil?
+  last_processed_date = Properties.get(LAST_PROCESSED_DATE_PROPERTY)
+  if last_processed_date.nil?
+    CDO.log.error "--start-date is required on first run of the script"
+    return
+  end
+  options[:start_date] = Date.parse(last_processed_date) + 1
+end
+
+if options[:end_date].nil?
+  options[:end_date] = Date.today - options[:retention_days]
+end
+
+# Options should not change from here on
+options.freeze
+CDO.log.info "Final calculated options: #{options}"
+
+# Main loop - Conceptually, we want to remove names from all rows with a
+# finished_at time between the start date and end date. We iterate over the
+# rows with a set of nested loops. The outer loop processes one day at a time,
+# checkpointing after each iteration by updating LAST_PROCESSED_DATE_PROPERTY
+# after each iteration. The inner loop processes the rows for a given day in
+# small batches to minimize impact to the database.
+date = options[:start_date]
+total_rows_processed = 0
+total_rows_updated = 0
+while date < options[:end_date]
+  CDO.log.info "Processing rows with finished_at=#{date}"
+
+  batch_start_time = date.to_time
+  batch_start_id = 0
+  batch_end_time = (date + 1).to_time
+  batch_number = 0
+  batch_rows_updated = 0
+
+  while batch_start_time < batch_end_time
+    ids_to_cleanup = []
+    last_row = nil
+
+    # Construct the sql where clauses for the start and end of the interval for
+    # this batch. Rows are ordered by (finished_at, id). The interval starts at
+    # (batch_start_time, batch_start_id) inclusive and ends at (batch_end_time, 0)
+    # exclusive.
+    interval_start_where_clause = Sequel.lit(
+      '(finished_at = ? AND id >= ?) OR finished_at > ?',
+      batch_start_time, batch_start_id, batch_start_time
+    )
+    interval_end_where_clause = Sequel.lit(
+      'finished_at < ?',
+      batch_end_time
+    )
+
+    HOC_ACTIVITY.select(:id, :name, :finished_at).
+      where(interval_start_where_clause).
+      where(interval_end_where_clause).
+      order(:finished_at, :id).
+      first(options[:batch_size]).
+      each do |row|
+        unless row[:name].nil?
+          ids_to_cleanup << row[:id]
+        end
+        total_rows_processed += 1
+        last_row = row
+      end
+
+    # If no rows were returned, we have processed all rows for this day
+    break if last_row.nil?
+
+    # Wrap the update in a transaction so we can rollback the changes on a dry run
+    DB.transaction do
```

Comment:
Do we need to carry out the updates in batches? Is there a functional or performance reason to have all updates in a batch succeed or fail as a unit? In practice, the number of rows this job needs to clear out will be roughly the number of certificates we issue each day (since the job runs daily).

Separate thought ... during Hour of Code, there’s an rapid increase in the number of certificates created each day. We should probably optimize performance for these daily numbers.

Reply:
I did try to design this to work for HoC days.  The batching is intended to avoid _lots_ of tiny transactions while still keeping individual transactions small.  The script iterates day by day and then, within each day, processes row in batches of `batch_size` which defaults to 100.  On busier days, we will need to process more batches, but each individual transaction will still read and write no more than 100 rows.  Does that help alleviate your concerns?

Reply:
The database performs reasonably well at runtime inserting and updating individual rows in this table as students carry out various activities that are persisted to it. If we don’t have any application logic that needs a group of rows to be successfully updated together or to all be rolled back if a single row can’t be updated, maybe we should just iterate over rows individually and update each one on its own. I don’t know whether there are performance benefits to updating batches vs updating individual rows. For the one-time historical wiping, batch updates seemed likely to improve performance, particularly to reduce excessive client/service interaction. We could experiment with setting the batch size to 1 if there are performance issues.